src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.CompilationTrackerState.cs

```diff
@@ -99,71 +135,95 @@ public WithCompilationTrackerState(Compilation compilationWithoutGeneratedDocume
+                    : base(isFrozen, compilationWithoutGeneratedDocuments, generatorInfo)
+                {
+                    // If we're not in the frozen state, the documents must *not* be final.  We're not a FinalState, and
+                    // as such, we must still determine what the up to date generated docs are.
+                    Contract.ThrowIfTrue(!IsFrozen && generatorInfo.DocumentsAreFinal);
+
+                    StaleCompilationWithGeneratedDocuments = staleCompilationWithGeneratedDocuments;
+                }
+            }
+
             /// <summary>
             /// A state where we are holding onto a previously built compilation, and have a known set of transformations
             /// that could get us to a more final state.
             /// </summary>
-            private sealed class InProgressState : WithCompilationTrackerState
+            private sealed class InProgressState : NonFinalWithCompilationTrackerState
             {
                 /// <summary>
                 /// The list of changes that have happened since we last computed a compilation. The oldState corresponds to
                 /// the state of the project prior to the mutation.
                 /// </summary>
                 public ImmutableList<(ProjectState oldState, CompilationAndGeneratorDriverTranslationAction action)> IntermediateProjects { get; }
 
-                /// <summary>
-                /// The result of taking the original completed compilation that had generated documents and updating them by
-                /// apply the <see cref="CompilationAndGeneratorDriverTranslationAction" />; this is not a correct snapshot in that
-                /// the generators have not been rerun, but may be reusable if the generators are later found to give the
-                /// same output.
-                /// </summary>
-                public Compilation? CompilationWithGeneratedDocuments { get; }
-
                 public InProgressState(
+                    bool isFrozen,
                     Compilation compilationWithoutGeneratedDocuments,
                     CompilationTrackerGeneratorInfo generatorInfo,
-                    Compilation? compilationWithGeneratedDocuments,
+                    Compilation? staleCompilationWithGeneratedDocuments,
                     ImmutableList<(ProjectState state, CompilationAndGeneratorDriverTranslationAction action)> intermediateProjects)
-                    : base(compilationWithoutGeneratedDocuments,
-                           generatorInfo.WithDocumentsAreFinal(false)) // since we have a set of transformations to make, we'll always have to run generators again
+                    : base(isFrozen,
+                           compilationWithoutGeneratedDocuments,
+                           generatorInfo,
+                           staleCompilationWithGeneratedDocuments)
                 {
                     Contract.ThrowIfTrue(intermediateProjects is null);
```

Comment:
A bit strange we're asserting this given we have nullable reference types...

Reply:
sure.  but nrt isn't foolproof.  i didn't mind keeping this around.